song-python-sdk/overture_song/entities.py

```diff
@@ -0,0 +1,348 @@
+import json
+from typing import Any, Type
+
+from dataclasses import dataclass
+
+from overture_song.validation import Validatable
+from overture_song.utils import Builder
+from typing import List
+from dataclasses import _isdataclass, asdict
+from overture_song.utils import check_type, check_state
+from overture_song.validation import non_null
+
+
+class Entity(object):
+
+    def to_json(self):
+        return json.dumps(self.to_dict(), indent=4)
+
+    def to_dict(self):
+        if _isdataclass(self):
+            return asdict(self)
+        else:
+            raise NotImplemented("not implemented for non-dataclass object")
+
+    def __str__(self):
+        return self.to_json()
+
+
+@dataclass(frozen=False)
+class Metadata(Entity):
+    info: dict = None
+
+    def __post_init__(self):
+        self.info = {}
+
+    def set_info(self, key: str, value: Any):
+        self.info[key] = value
+
+    def add_info(self, data: dict):
+        if data is None:
+            return
+        self.info.update(data)
+
+
+@dataclass(frozen=False)
+class Study(Metadata, Validatable):
+    studyId: str = None
+    name: str = None
+    organization: str = None
+    description: str = None
+
+    def validate(self):
+        raise NotImplemented("not implemented")
+
+
+    @classmethod
+    def create(cls, studyId, name=None, description=None, organization=None):
+        s = Study()
+        s.studyId = studyId
+        s.name = name
+        s.description = description
+        s.organization = organization
+        return s
+
+
+    @classmethod
+    def create_from_raw(cls, study_obj):
+        return Study.create(
+            study_obj.studyId,
+            name=study_obj.name,
+            description=study_obj.description,
+            organization=study_obj.organization)
+
+
+@dataclass(frozen=False)
+class File(Metadata, Validatable):
+    objectId: str = None
+    analysisId: str = None
+    fileName: str = None
+    studyId: str = None
+    fileSize: int = -1
+    fileType: str = None
+    fileMd5sum: str = None
+    fileAccess: str = None
+
+    def validate(self):
+        raise NotImplemented("not implemented")
+
+    @classmethod
+    def create(cls,fileName, fileSize, fileType, fileMd5sum,
+               fileAccess, studyId=None, analysisId=None, objectId=None, info={}):
+        f = File()
+        f.objectId = objectId
+        f.analysisId = analysisId
+        f.studyId = studyId
+        f.fileType = fileType
+        f.fileSize = fileSize
+        f.info = info
+        f.fileMd5sum = fileMd5sum
+        f.fileAccess = fileAccess
+        f.fileName = fileName
+        return f
+
+
+
+@dataclass(frozen=False)
+class Sample(Metadata, Validatable):
+    sampleId: str = None
+    specimenId: str = None
+    sampleSubmitterId: str = None
```

Comment:
camelCase is being used on a lot of these

Reply:
CamelCase is required since the member variable names are used when automatically serializing the objects into json. By using camel case only for member variables of classes annotated with @dataclass, the code is simplified, more maintainable and cleaner